File: TreeTraversal.py


class tree(object):
    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data

class traverse(object):
    def __init__(self):
        print "Calling contructor"

    def pre_order(self, root):
        if root:
            print "%2d -> ",  root.data
            self.pre_order(root.left)
            self.pre_order(root.right)

    def post_order(self, root):
        if root == None:
            return
        self.post_order( root.left)
        self.post_order( root.right)
        print "%2d -> ",  root.data
    
    def in_order(self, root):
        if root == None:
            return
        self.in_order( root.left)
        print "%2d -> ",  root.data
        self.in_order( root.right)
#Main starts from here

# creating the tree
root = tree(5)
root.left = tree(2)
root.right = tree(7)
root.left.left = tree(9)
root.left.left.right = tree(4)
root.left.left.left = tree(3)
root.right.left = tree(8)
root.right.right = tree(1)
root.right.right.left = tree(6)

func__ = traverse()
print "Printing pre order"
func__.pre_order(root)
print "Printing post order"
func__.post_order(root)
print "Printing in order"
func__.in_order(root)

File: TreeTraversal.c
#include <stdio.h>
#include <stdlib.h>

typedef struct tree
{
    int data;
    struct tree *left;
    struct tree *right;
} *pDataTree, DataTree;

pDataTree getNode(int data)
{
    pDataTree buf = malloc(sizeof(struct tree));
    buf->left = NULL;
    buf->right = NULL;
    buf->data = data;
    return buf;
}

void print_preorder(pDataTree root)
{
    if(root == NULL )
        return;
    printf("%2d -> ", root->data);
    print_preorder(root->left);
    print_preorder(root->right);
}

void in_preorder(pDataTree root)
{
    if(root == NULL )
        return;
    print_preorder(root->left);
    printf("%2d -> ", root->data);
    print_preorder(root->right);
}

void post_preorder(pDataTree root)
{
    if(root == NULL )
        return;
    print_preorder(root->left);
    print_preorder(root->right);
    printf("%2d -> ", root->data);
}
void Create_Node(pDataTree *root)
{
    *root = getNode(5);
    (*root)->left = getNode(2);
    (*root)->right = getNode(7);
    (*root)->left->left = getNode(9);
    (*root)->left->left->left = getNode(3);
    (*root)->left->left->right = getNode(4);

    (*root)->right->left = getNode(8);
    (*root)->right->right = getNode(1);
    (*root)->right->right->left = getNode(6);
}

int main(int argc, char * argv[])
{
    pDataTree root = NULL;
    Create_Node(&root); 
    printf("Printing Preorder\n");
    print_preorder(root);
    printf("Printing Inorder\n");
    in_preorder(root);
    printf("Printing PostOrder\n");
    post_preorder(root);
    printf("NULL\n");
    return 0;
}
